### PR TITLE
Rework npm run network-audit into a browser test suite

### PR DESCRIPTION
**Pull request by mariospr**
This PR creates a new browser tests suite `brave_network_audit_tests` that will
perform similar checks, but in a more reliable and flexible way (i.e. will allow
to control navigation events and avoid JSON parsing-related errors on CI).

Such test suite will now be run via the `npm run test-security` script with
appropriate values to extend the tests' default timeout, in order to accomodate
the fact that these tests will simulate leaving the browser open for about 2
minutes, in order to gather enough information and make sure that only allowed
network requests are issued during that time.

This PR also removes Google-owned prefixes and patterns from network audit
whitelist since these entries correspond with internal 307 responses that should
not be going out to the network. Last, all the whitelisted values are now moved
into //brave/browser/net/brave_network_audit_whitelists.h so that they are all
in a single placed owned by the security team, as set in .github/CODEOWNERS.

Resolves https://github.com/brave/brave-browser/issues/7207
Resolves https://github.com/brave/brave-browser/issues/7281

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one --> Approved in https://github.com/brave/security/issues/611
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

N/A

### CODE REVIEW

**Comment by mariospr** [comment]
@mihaiplesa I added you explicitly as a reviewer mainly because of the "Remove all trace of npm run network-audit and replace it where needed" change, which touches some Jenkins-related files and I'm not fully sure about those changes. PTAL, thanks!

File: Jenkinsfile [hunk 1]
```diff
@@ -47,15 +47,13 @@ pipeline {
                         CHANGE_BRANCH_ENCODED = java.net.URLEncoder.encode(CHANGE_BRANCH, 'UTF-8')
                         def prDetails = readJSON(text: httpRequest(url: GITHUB_API + '/' + REPO + '/pulls?head=brave:' + CHANGE_BRANCH_ENCODED, customHeaders: GITHUB_AUTH_HEADERS, quiet: true).content)[0]
                         SKIP = prDetails.draft.equals(true) || prDetails.labels.count { label -> label.name.equalsIgnoreCase('CI/skip') }.equals(1) || prDetails.labels.count { label -> label.name.equalsIgnoreCase("CI/skip-${PLATFORM}") }.equals(1)
-                        RUN_NETWORK_AUDIT = prDetails.labels.count { label -> label.name.equalsIgnoreCase('CI/run-network-audit') }.equals(1)
```
Comment (mariospr) [inline Jenkinsfile:50]: @mihaiplesa Based on our discussion on Slack, with further clarification from @bridiver, I think this change it's not needed. Looks like what needs to be changes should happen in Jenkins itself, which it's what I think https://github.com/brave/devops/issues/5965 is for.

Removing this and force-pushing

File: browser/net/brave_network_audit_whitelists.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+// Before adding to this list, get approval from the security team.
+constexpr const char* kWhitelistedUrlProtocols[] = {
+    "chrome-extension", "chrome", "brave", "file", "data", "blob",
+};
+
+// Before adding to this list, get approval from the security team.
+constexpr const char* kWhitelistedUrlPrefixes[] = {
+    // allowed because it 307's to https://componentupdater.brave.com
+    "http://componentupdater.brave.com/service/update2",
```
Comment (diracdeltas) [inline browser/net/brave_network_audit_whitelists.h:21]: this http:// version is probably fine to remove too

File: browser/net/brave_network_audit_browsertest.cc [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+// Loads brave://welcome first to simulate a first run and then loads another
+// URL, waiting some time after each load to allow gathering network requests.
+IN_PROC_BROWSER_TEST_F(BraveNetworkAuditTest, BasicTests) {
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), GURL("brave://welcome")));
+  WaitForTimeout(kMaxTimeoutPerLoadedURL);
+
+  ASSERT_TRUE(embedded_test_server()->Start());
+  GURL simple_url(embedded_test_server()->GetURL("/simple.html"));
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), simple_url));
```
Comment (diracdeltas) [inline browser/net/brave_network_audit_browsertest.cc:243]: this is great; i think eventually we should load `brave://rewards`, simulate turning that on, then wait for any more requests. could be done as part of this PR or as a follow up.

**Comment by mariospr** [comment]
@bridiver @iefremov Security review passed in pproved in https://github.com/brave/security/issues/611, PTAL when you have a moment, thanks!

**Comment by mariospr** [comment]
> @bridiver @iefremov Security review passed in [brave/security#611](https://github.com/brave/security/issues/611), PTAL when you have a moment, thanks!

@bridiver @iefremov Kindly pinging you again :innocent:

File: browser/net/brave_network_audit_browsertest.cc [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+ public:
+  BraveNetworkAuditTest() = default;
+
+  void SetUpOnMainThread() override {
+    InProcessBrowserTest::SetUpOnMainThread();
+
+    ASSERT_TRUE(embedded_test_server()->Start());
+
+    // Create and start the Rewards service
+    rewards_service_ = static_cast<brave_rewards::RewardsServiceImpl*>(
```
[the remark this answers is not in the document]
Comment (diracdeltas) [reply browser/net/brave_network_audit_browsertest.cc:181]: yes, in addition to testing for unexpected startup requests + requests upon rewards enabled, we will eventually want to check other things like:

* requests from Tor windows only go through Tor
* safebrowsing-triggered requests do not directly connect to Google
* other potential network privacy leaks that come up often

Marshall also asked if this test could check that talk.brave.com does not make unexpected connections to Jitsi, but that is better suited for a separate test suite for the brave talk repo.

**Comment by mariospr** [comment]
In the end I managed to get ahold of a MacPro machine (Big Sur / 16 cores @ 3.2Ghz / 96GB DDR4) and I couldn't reproduce this issue: I built with  `npm run build -- Release` first and then I run `npm run network-audit -- Release` and it did finished compiling & linking in ~15 min (linking took quite a big % of the time), which is way faster than the ~1h observed in CI.

I've also talked to @iefremov who run a similar test in his local MacPro and couldn't reproduce the issue either, so I'm afraid this is somehow related to something specific to the CI bot, and might need some debugging related to that.

@wknapik WDYT?

**Comment by wknapik** [comment]
We have two types of macOS nodes - minis and imacs. Triggered a clean build on an imac. Let's see if the extra resources change anything.

**Comment by wknapik** [comment]
On an imac, it took 46min to build/run audit-network ([log](https://ci.brave.com/job/pr-brave-browser-issues-7207-macos/25/execution/node/232/log/)).

What's the conclusion here? That our mac nodes don't have enough hardware resources?

**Comment by mariospr** [comment]
> On an imac, it took 46min to build/run audit-network ([log](https://ci.brave.com/job/pr-brave-browser-issues-7207-macos/25/execution/node/232/log/)).

Yes, I was now looking into the logs and found that too. In more detail:

The **build took about 35 min** (still a lot, but ~30 min less than before):
```
17:25:43  > ninja -C /<path/to/workspace>/src/out/Release brave/test:brave_network_audit_tests
17:25:43  ninja: Entering directory `/<path/to/workspace>/src/out/Release'
17:25:55  [1/5064] ACTION //media:media_buildflags(//build/toolchain/mac:clang_x64)
17:25:55  [2/5064] ACTION //brave/test:brave_network_audit_tests__runner(//build/toolchain/mac:clang_x64)
17:25:55  [3/5064] STAMP obj/media/media_buildflags.stamp

[...]

17:50:55  [5009/5010] AR obj/chrome/test/libtest_support.a
17:57:09  [5010/5010] LINK ./brave_network_audit_tests
17:57:09  using git cache path /Users/jenkins/cache
17:57:09  -------------------------------
```
...and then the **test itself took again about 15min**, as expected:
```
17:57:09  > /<path/to/workspace>/src/out/Release/brave_network_audit_tests --enable-logging=stderr --v=0 --test-launcher-jobs=4 --ui-test-action-timeout=330000 --test-launcher-timeout=1000000 Release --output brave_network_audit_tests.xml --gtest_output=xml:brave_network_audit_tests.xml
17:57:09  IMPORTANT DEBUGGING NOTE: each test is run inside its own process.
17:57:09  For debugging a test inside a debugger, use the
17:57:09  --gtest_filter=<your_test_name> flag along with either
17:57:09  --single-process-tests (to run the test in one launcher/browser process) or
17:57:09  --single-process (to do the above, and also run Chrome in single-process mode).
17:57:09  Using sharding settings from environment. This is shard 0/1
17:57:09  Using 2 parallel jobs.
17:57:09  [ RUN      ] BraveNetworkAuditTest.BasicTests

[...]

18:12:06  [2/2] BraveNetworkAuditTest.BasicTests (902722 ms)
18:12:06  SUCCESS: all tests passed.
```

> What's the conclusion here? That our mac nodes don't have enough hardware resources?

Sadly, I think that must be it, yes. Would it be possible to have the CI switch to imacs?

**Comment by wknapik** [comment]
Triggered another rebuild on the same imac to see more build/run times.

**Comment by iefremov** [comment]
we could trigger another one on a Mini, just to ensure that that was not a glitch :) 
I'm surprised we run into this problem, because building these tests shouldn't be super different than normal tests builds

**Comment by wknapik** [comment]
In another rebuild on the same imac network audit took 57min to build/run ([log](https://ci.brave.com/job/pr-brave-browser-issues-7207-macos/27/execution/node/232/log/?consoleFull)).

**Comment by wknapik** [comment]
In a build on the imac with 384G ram network audit took 45min to build and run ([log](https://ci.brave.com/job/pr-brave-browser-issues-7207-macos/30/execution/node/232/log/?consoleFull)).

**Comment by wknapik** [comment]
Build + execution times for network audit:
* Linux: 20min
* Windows: 27min
* macOS
  * mac mini (used in PRs along with imacs): 80min, ?min (still running)
  * imac (used in public pipelines): 46min, 57min, 45min (with 384G ram), 47min

**Comment by mariospr** [comment]
> While performance on Linux is fine, Windows isn't great and macOS is bad.
> 
> On macOS this appears to be tied to available cpu resources. For now, this is what we have.
> 
> Do we want to merge and enable on all platforms given the numbers above?

@iefremov @mihaiplesa @diracdeltas WDYT?

FWIW; the last run on CI failed on Mac but it looks like an unrelated problem:
https://ci.brave.com/job/pr-brave-browser-issues-7207-macos/32/execution/node/253/log/

**Comment by bsclifton** [comment]
20 and 27 mins not terrible; the macOS ones - yikes 😱 We always could just start with Linux/Windows first until time (on macOS) is comparable (in 20 min range)

**Comment by wknapik** [comment]
In PRs network audit runs only on demand, so I don't think we need to worry about that.

As for public pipelines - @diracdeltas how likely would it be for this network audit to find nothing on Linux and Windows and something on macOS?

It feels wrong not to run this on macOS, but our macOS/iOS resources are already scarce, so I'd be hesitant to make the situation worse by extending the runtime for macOS pipelines by 45-60min...

**Comment by diracdeltas** [comment]
> how likely would it be for this network audit to find nothing on Linux and Windows and something on macOS?

I don't think this has happened before but it's somewhat likely to happen given that the networking code in Chromium depends on the OS in many places.

Perhaps for now this could be run every 3-4 macOS builds at random if the runtime is a concern?

**Review by mihaiplesa — APPROVED** [review]

**Comment by wknapik** [comment]
Decided with @mihaiplesa to run network audit in public pipelines on macos only on the nightly channel. We can expand to other channels if this doesn't cause a shortage of mac executors. Will push the devops changes to reflect that shortly.

**Review by wknapik — APPROVED** [review]

**Pull request merged**